bpf/custom/bpf_custom.c

```diff
@@ -0,0 +1,34 @@
+// SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
+/* Copyright (C) 2020 Authors of Cilium */
+
+#include <bpf/ctx/skb.h>
+#include <bpf/api.h>
+
+#include "lib/common.h"
+
+#define DIRECTION_INGRESS	0
+#define DIRECTION_EGRESS	1
+
+/* Include custom header file (.h) containing the implementation for
+ * custom_prog(), for example:
+ *
+ *     #include "bytecount.h"
+ */
+#include "bytecount.h"
+
+__section("custom")
+int custom_hook(const struct __ctx_buff *ctx)
+{
+	__u32 custom_meta = ctx_load_meta(ctx, CB_CUSTOM_CALLS);
+	int direction = (custom_meta >> 31) & 0x1;
+	__u32 identity = custom_meta & 0xffffff;
+	int ret = (custom_meta >> 24) & 0xff;
+
+	/* Call user-defined function from custom header file. */
+	custom_prog(ctx, identity, direction);
```

Comment:
If the custom program called certain helper functions (I'm thinking like `bpf_redirect()`), then it would influence the path that the `skb` will take. How much guardrails do we want to put in place for cases like this?

Reply:
During the last #sig-datapath meeting, we discussed using `const` (including in helpers) to have at least a compile-time warning against that.

Reply:
I don't see an easy way to prevent such scenario. We indeed discussed making the `ctx` pointer `const` to discourage modifications of the packet. But ultimately, we have no guarantee that the users will use the `bpf_custom.c` “landing point” for the custom hook, they can as well build and load their own programs outside of the Cilium repo.

I don't think we should look for ways to strongly prevent that, but instead this should be correctly documented (TBD). We should make very clear that modifying the packet or the return value at this stage has a huge chance of crashing the datapath. But if we can make sure that people are aware of that and they prefer to experiment and live on the edge, I think we should not make it a fight to load custom-custom programs.